hathor_tests/nanocontracts/test_blueprints/hathordice.py

```diff
@@ -0,0 +1,284 @@
+            self.liquidity_providers[ctx.caller_id] = adjusted_amount
+        else:
+            self.liquidity_providers[ctx.caller_id] += adjusted_amount
+
+        self.total_liquidity_provided += adjusted_amount
+        self.available_tokens += amount
+
+        return adjusted_amount
+
+    @view
+    def calculate_adjusted_liquidity(self, amount: Amount) -> int:
+        """Calculate adjusted liquidity when ADDING liquidity (floor division)."""
+        # x = amount
+        # y = adjusted_amount
+        # A = self.available_tokens
+        # L = self.total_liquidity_provided
+        #
+        # x = y * (A + x) / (L + y)
+        # x * (L + y) = y * (A + x)
+        # xL + xy = Ay + xy
+        # xL = Ay
+        # y = x * L // A
+        if self.available_tokens == 0:
+            return amount
+        return (amount * self.total_liquidity_provided) // self.available_tokens
+
+    @view
+    def calculate_adjusted_liquidity_removal(self, amount: Amount) -> int:
+        """Calculate adjusted liquidity when REMOVING liquidity (ceiling division for security)."""
+        if self.available_tokens == 0:
+            return amount
+        return self._ceil_div(
+            amount * self.total_liquidity_provided,
+            self.available_tokens
+        )
+
+    @public(allow_withdrawal=True)
+    def remove_liquidity(self, ctx: Context) -> int:
+        action = self._get_action(ctx, NCWithdrawalAction)
+
+        allowed_withdrawal = self.calculate_address_maximum_liquidity_removal(ctx.caller_id)
+        if action.amount > allowed_withdrawal:
+            raise NCFail('too large withdrawal')
+
+        amount = action.amount
+        adjusted_amount = self.calculate_adjusted_liquidity_removal(amount)
+
+        self.liquidity_providers[ctx.caller_id] -= adjusted_amount
+        self.total_liquidity_provided -= adjusted_amount
+        self.available_tokens -= amount
+
+        return adjusted_amount
+
+    @view
+    def calculate_maximum_liquidity_removal(self, amount: Amount) -> Amount:
+        if self.total_liquidity_provided == 0:
+            return 0
+        assert amount >= 0
+        return (self.available_tokens * amount) // self.total_liquidity_provided
+
+    @view
+    def calculate_address_maximum_liquidity_removal(self, caller_id: CallerId) -> Amount:
+        amount = self.liquidity_providers.get(caller_id, 0)
+        assert amount >= 0
+        return self.calculate_maximum_liquidity_removal(amount)
+
+    @public(allow_deposit=True)
+    def place_bet(self, ctx: Context, bet_amount: Amount, threshold: int) -> int:
+        if not isinstance(ctx.caller_id, Address):
+            raise NCFail('only users are allowed')
+        if bet_amount <= 0:
+            raise NCFail('bet amount must be positive')
+        if bet_amount > self.max_bet_amount:
+            raise NCFail('bet amount is too high')
+
+        if threshold < 0:
+            raise NCFail('threshold must be positive')
```

Comment:
What about `0`? Should it be `non-negative` or strictly `positive`, if it must be `positive` then the check should be `<= 0`

Reply:
Fixed. Thanks!